File: apps/ios/RemoteCodex/Core/ThreadProjection/ThreadOptimisticProjection.swift
import Foundation

struct OptimisticPromptTurn: Equatable {
    var id: String
    var serverTurnId: String?
    var prompt: String
    var startedAt: String
    var model: String?
    var status: OptimisticPromptStatus = .sending
    var error: String?

    func withAcceptedThread(_ thread: SupervisorThreadSummary) -> OptimisticPromptTurn {
        var next = self
        next.model = thread.model ?? model
        next.status = .inProgress
        next.error = nil
        return next
    }

    func withStartedTurn(_ turnId: String) -> OptimisticPromptTurn {
        guard serverTurnId == nil else { return self }
        var next = self
        next.serverTurnId = turnId
        next.status = .inProgress
        return next
    }

    func withFailure(_ message: String) -> OptimisticPromptTurn {
        var next = self
        next.status = .failed
        next.error = message
        return next
    }
}

enum OptimisticPromptStatus: Equatable {
    case sending
    case inProgress
    case failed
}

func shouldClearOptimisticPrompt(detail: SupervisorThreadDetail, optimistic: OptimisticPromptTurn?) -> Bool {
    guard let optimistic, optimistic.status != .failed else { return false }
    return detail.turns.contains { turn in
        turn.matches(optimistic: optimistic) && turn.hasMatchingUserMessage(optimistic.prompt)
    }
}

func applyOptimisticPromptProjection(
    detail: SupervisorThreadDetail,
    optimistic: OptimisticPromptTurn?
) -> SupervisorThreadDetail {
    guard let optimistic else { return detail }

    let matchingIndex = detail.turns.firstIndex { $0.matches(optimistic: optimistic) }
    let optimisticTurn = optimistic.toSupervisorTurn(indexFallback: detail.turns.count)
    let projectedTurns: [SupervisorThreadTurn] = if let matchingIndex {
        detail.turns.enumerated().map { index, turn in
            index == matchingIndex ? turn.withOptimisticPromptItems(optimistic) : turn
        }
    } else {
        detail.turns + [optimisticTurn]
    }

    let running = optimistic.status != .failed
    var next = detail
    next.turns = projectedTurns
    next.turnCount = max(detail.turnCount ?? 0, projectedTurns.count)
    next.totalTurnCount = max(detail.totalTurnCount ?? 0, projectedTurns.count)
    if running {
        next.liveItemCount = max(detail.liveItemCount ?? 0, 1)
        next.thread.status = "running"
        next.thread.updatedAt = optimistic.startedAt
    }
    return next
}

private extension SupervisorThreadTurn {
    func matches(optimistic: OptimisticPromptTurn) -> Bool {
        id == optimistic.serverTurnId ||
            id == optimistic.id ||
            hasMatchingUserMessage(optimistic.prompt) ||
            hasPhotoPromptMatch(optimistic.prompt)
    }

    func hasMatchingUserMessage(_ prompt: String) -> Bool {
        let normalizedPrompt = prompt.normalizedPromptTextForProjection
        return items.contains { item in
            item.kind == "userMessage" &&
                item.text?.normalizedPromptTextForProjection == normalizedPrompt
        }
    }

    func hasPhotoPromptMatch(_ prompt: String) -> Bool {
        guard prompt.contains("[PHOTO ") else { return false }
        return items.contains { item in
            item.kind == "userMessage" &&
                (item.text?.contains("[PHOTO ") == true || item.text?.localizedCaseInsensitiveContains("![") == true)
        }
    }

    func withOptimisticPromptItems(_ optimistic: OptimisticPromptTurn) -> SupervisorThreadTurn {
        let userMessage = SupervisorThreadTurnItem(
            id: "\(optimistic.id)-user-message",
            kind: "userMessage",
            text: optimistic.prompt,
            status: nil,
            sequence: nil,
            callId: nil,
            toolName: nil,
            payload: nil
        )
        let orderedItems = hasMatchingUserMessage(optimistic.prompt) || hasPhotoPromptMatch(optimistic.prompt)
            ? items
            : [userMessage] + items
        let withPlaceholder = optimistic.status != .failed &&
            !orderedItems.contains { $0.kind == "agentMessage" }
            ? orderedItems + [optimistic.assistantPlaceholder(indexFallback: orderedItems.count + 1)]
            : orderedItems
        var next = self
        next.status = switch optimistic.status {
        case .failed:
            "failed"
        case .sending, .inProgress:
            status == "completed" || status == "failed" ? status : "running"
        }
        next.error = optimistic.error ?? error
        next.items = withPlaceholder.sortedOptimisticItems()
        return next
    }
}

private extension OptimisticPromptTurn {
    func toSupervisorTurn(indexFallback: Int) -> SupervisorThreadTurn {
        SupervisorThreadTurn(
            id: serverTurnId ?? id,
            startedAt: startedAt,
            status: status == .failed ? "failed" : "running",
            error: error,
            model: model,
            tokenUsage: nil,
            items: [
                SupervisorThreadTurnItem(
                    id: "\(id)-user-message",
                    kind: "userMessage",
                    text: prompt,
                    status: nil,
                    sequence: nil,
                    callId: nil,
                    toolName: nil,
                    payload: nil
                ),
                assistantPlaceholder(indexFallback: indexFallback + 1)
            ]
        )
    }

    func assistantPlaceholder(indexFallback: Int) -> SupervisorThreadTurnItem {
        SupervisorThreadTurnItem(
            id: "\(id)-assistant-placeholder",
            kind: "agentMessage",
            text: "",
            status: status == .failed ? nil : "running",
            sequence: indexFallback,
            callId: nil,
            toolName: nil,
            payload: nil
        )
    }
}

private extension [SupervisorThreadTurnItem] {
    func sortedOptimisticItems() -> [SupervisorThreadTurnItem] {
        enumerated()
            .sorted { lhs, rhs in
                let leftGroup = lhs.element.kind == "userMessage" ? 0 : 1
                let rightGroup = rhs.element.kind == "userMessage" ? 0 : 1
                if leftGroup != rightGroup {
                    return leftGroup < rightGroup
                }
                let leftSequence = lhs.element.sequence ?? Int.max
                let rightSequence = rhs.element.sequence ?? Int.max
                if leftSequence != rightSequence {
                    return leftSequence < rightSequence
                }
                return lhs.offset < rhs.offset
            }
            .map(\.element)
    }
}

private extension String {
    var normalizedPromptTextForProjection: String {
        trimmingCharacters(in: .whitespacesAndNewlines)
            .split(whereSeparator: \.isWhitespace)
            .joined(separator: " ")
    }
}

File: apps/android/app/src/main/java/com/remotecodex/android/thread/ThreadOptimisticProjection.kt
package com.remotecodex.android.thread

import com.remotecodex.android.api.SupervisorThreadDetail
import com.remotecodex.android.api.SupervisorThreadSummary
import com.remotecodex.android.api.SupervisorThreadTurn
import com.remotecodex.android.api.SupervisorThreadTurnItem

data class OptimisticPromptTurn(
    val id: String,
    val serverTurnId: String? = null,
    val prompt: String,
    val startedAt: String,
    val model: String?,
    val status: OptimisticPromptStatus = OptimisticPromptStatus.Sending,
    val error: String? = null,
)

enum class OptimisticPromptStatus {
    Sending,
    InProgress,
    Failed,
}

fun OptimisticPromptTurn.withAcceptedThread(thread: SupervisorThreadSummary): OptimisticPromptTurn {
    return copy(
        model = thread.model ?: model,
        status = OptimisticPromptStatus.InProgress,
        error = null,
    )
}

fun OptimisticPromptTurn.withStartedTurn(turnId: String): OptimisticPromptTurn {
    return if (serverTurnId == null) {
        copy(serverTurnId = turnId, status = OptimisticPromptStatus.InProgress)
    } else {
        this
    }
}

fun OptimisticPromptTurn.withFailure(message: String): OptimisticPromptTurn {
    return copy(status = OptimisticPromptStatus.Failed, error = message)
}

fun shouldClearOptimisticPrompt(
    detail: SupervisorThreadDetail,
    optimistic: OptimisticPromptTurn?,
): Boolean {
    if (optimistic == null || optimistic.status == OptimisticPromptStatus.Failed) {
        return false
    }
    return detail.turns.any { turn -> turn.matchesOptimisticPrompt(optimistic) && turn.hasMatchingUserMessage(optimistic.prompt) }
}

fun applyOptimisticPromptProjection(
    detail: SupervisorThreadDetail,
    optimistic: OptimisticPromptTurn?,
): SupervisorThreadDetail {
    if (optimistic == null) {
        return detail
    }

    val matchingIndex = detail.turns.indexOfFirst { turn -> turn.matchesOptimisticPrompt(optimistic) }
    val optimisticTurn = optimistic.toSupervisorTurn(indexFallback = detail.turns.size)
    val projectedTurns = if (matchingIndex >= 0) {
        detail.turns.mapIndexed { index, turn ->
            if (index == matchingIndex) {
                turn.withOptimisticPromptItems(optimistic)
            } else {
                turn
            }
        }
    } else {
        detail.turns + optimisticTurn
    }

    val running = optimistic.status != OptimisticPromptStatus.Failed
    return detail.copy(
        turns = projectedTurns,
        turnCount = maxOf(detail.turnCount, projectedTurns.size),
        totalTurnCount = maxOf(detail.totalTurnCount, projectedTurns.size),
        liveItemCount = if (running) maxOf(detail.liveItemCount, 1) else detail.liveItemCount,
        thread = if (running) {
            detail.thread.copy(status = "running", updatedAt = optimistic.startedAt)
        } else {
            detail.thread
        },
    )
}

private fun SupervisorThreadTurn.matchesOptimisticPrompt(optimistic: OptimisticPromptTurn): Boolean {
    return id == optimistic.serverTurnId ||
        id == optimistic.id ||
        hasMatchingUserMessage(optimistic.prompt) ||
        hasPhotoPromptMatch(optimistic.prompt)
}

private fun SupervisorThreadTurn.hasMatchingUserMessage(prompt: String): Boolean {
    val normalizedPrompt = prompt.normalizedPromptText()
    return items.any { item ->
        item.kind == "userMessage" &&
            item.text.normalizedPromptText() == normalizedPrompt
    }
}

private fun SupervisorThreadTurn.hasPhotoPromptMatch(prompt: String): Boolean {
    if (!prompt.contains("[PHOTO ")) {
        return false
    }
    return items.any { item ->
        item.kind == "userMessage" &&
            (item.text.contains("[PHOTO ") || item.text.contains("![", ignoreCase = true))
    }
}

private fun SupervisorThreadTurn.withOptimisticPromptItems(
    optimistic: OptimisticPromptTurn,
): SupervisorThreadTurn {
    val userMessage = SupervisorThreadTurnItem(
        id = "${optimistic.id}-user-message",
        kind = "userMessage",
        text = optimistic.prompt,
    )
    val hasUser = hasMatchingUserMessage(optimistic.prompt) || hasPhotoPromptMatch(optimistic.prompt)
    val orderedItems = if (hasUser) {
        items
    } else {
        listOf(userMessage) + items
    }
    val withPlaceholder = if (
        optimistic.status != OptimisticPromptStatus.Failed &&
        orderedItems.none { item -> item.kind == "agentMessage" }
    ) {
        orderedItems + SupervisorThreadTurnItem(
            id = "${optimistic.id}-assistant-placeholder",
            kind = "agentMessage",
            text = "",
            status = "running",
            sequence = Int.MAX_VALUE,
        )
    } else {
        orderedItems
    }
    return copy(
        status = when (optimistic.status) {
            OptimisticPromptStatus.Failed -> "failed"
            OptimisticPromptStatus.Sending,
            OptimisticPromptStatus.InProgress,
            -> if (status == "completed" || status == "failed") status else "running"
        },
        error = optimistic.error ?: error,
        items = withPlaceholder.sortOptimisticItems(),
    )
}

private fun OptimisticPromptTurn.toSupervisorTurn(indexFallback: Int): SupervisorThreadTurn {
    val turnId = serverTurnId ?: id
    val statusText = when (status) {
        OptimisticPromptStatus.Failed -> "failed"
        OptimisticPromptStatus.Sending,
        OptimisticPromptStatus.InProgress,
        -> "running"
    }
    return SupervisorThreadTurn(
        id = turnId,
        startedAt = startedAt,
        status = statusText,
        error = error,
        model = model,
        tokenUsage = null,
        items = listOf(
            SupervisorThreadTurnItem(
                id = "$id-user-message",
                kind = "userMessage",
                text = prompt,
            ),
            SupervisorThreadTurnItem(
                id = "$id-assistant-placeholder",
                kind = "agentMessage",
                text = "",
                status = if (status == OptimisticPromptStatus.Failed) null else "running",
                sequence = indexFallback + 1,
            ),
        ),
    )
}

private fun List<SupervisorThreadTurnItem>.sortOptimisticItems(): List<SupervisorThreadTurnItem> {
    return mapIndexed { index, item -> index to item }
        .sortedWith(
            compareBy<Pair<Int, SupervisorThreadTurnItem>>(
                { (_, item) -> if (item.kind == "userMessage") 0 else 1 },
                { (_, item) -> item.sequence ?: Int.MAX_VALUE },
                { (index, _) -> index },
            ),
        )
        .map { (_, item) -> item }
}

private fun String.normalizedPromptText(): String {
    return trim().replace(Regex("\\s+"), " ")
}
